crates/bevy_ecs/src/query/iter.rs

```diff
@@ -650,8 +648,7 @@ impl<'w, 's, D: ReadOnlyQueryData, F: QueryFilter, const K: usize> FusedIterator
 }
 
 struct QueryIterationCursor<'w, 's, D: QueryData, F: QueryFilter> {
-    table_id_iter: std::slice::Iter<'s, TableId>,
-    archetype_id_iter: std::slice::Iter<'s, ArchetypeId>,
+    storage_id_iter: std::slice::Iter<'s, StorageId>,
```

Comment:
Does that mean that the `table_entities` and `archetype_entities` below could have a similar optiimization as `storage_entities: &'w [StorageEntities]`?

Reply:
This is what #5085 does, however the benefit may be limited as this struct doesn't really exist at runtime: it's never formally materialized to the stack or the heap under normal use cases and any inlined iteration will decompose the fetches and updates.

Compare this with the Vec in QueryState, which requires both stack and heap space due to being a persisted heap allocated backing for Query. There's real memory savings by using the union.

Doable but I'm not sure if the savings are worth the introduction of more unsafe and readability impact.